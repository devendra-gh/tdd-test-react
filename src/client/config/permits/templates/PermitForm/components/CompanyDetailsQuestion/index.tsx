import React, { useState } from 'react';
import { IVariables } from '@tamm/app-composer';
import FormTemplate from '@tamm/ui-lib-v2-form-template';
import {
  checkValidationField,
  validateSharePercentage,
  validationTypes,
} from 'client/config/utils/checkValidation';
import {
  companies,
  nationalities,
  emirates,
  legalForms,
  representativeType,
  countries,
} from '../../../../utils/constants';

const margin = {
  margin: '20px',
};

export const checkLicenseDetails = async (
  licenseNumber: string,
  fetchLicenseDetails: any,
  callback: any,
) => {
  const details = await fetchLicenseDetails(licenseNumber);
  if (details) {
    const location = details.businessLicCity || '';
    const businessNameEn =
      details.businessNameEng || details.businessNameArb || '';
    const businessNameAr =
      details.businessNameArb || details.businessNameEng || '';

    callback({
      licenseValid: true,
      location,
      businessNameEn,
      businessNameAr,
    });
  } else {
    callback({
      licenseValid: false,
      location: '',
      businessNameEn: '',
      businessNameAr: '',
    });
  }
};

const CompanyDetailsQuestion = ({
  values,
  handleChange,
  handleCompanyTypeChange,
  handleRepresentativeTypeChange,
  companyType,
  companyTypes,
  serviceType,
  fetchLicenseDetails,
  startShowingErrors,
  i18n,
}: {
  values: IVariables;
  handleChange: Function;
  handleCompanyTypeChange: Function;
  handleRepresentativeTypeChange: Function;
  companyType: string;
  companyTypes: any[];
  serviceType: string;
  fetchLicenseDetails: Function;
  startShowingErrors: boolean;
  i18n: Function;
}) => {
  if (!serviceType) {
    return null;
  }

  const [loading, setLoading] = useState(false);

  const onBlurLicenseNumber = async () => {
    const typeValidation = checkValidationField(
      {
        type: validationTypes.CN_NUMBER,
      },
      values.licenseNo || null,
      true,
      i18n,
    );
    if (!typeValidation && values.licenseNo) {
      setLoading(true);
      await checkLicenseDetails(
        values.licenseNo,
        fetchLicenseDetails,
        handleChange,
      );
      setLoading(false);
    }
  };

  // License field for onshore comapnies
  const licenceNumberField = {
    label: i18n('field.comLicenseNumber'),
    value: values.licenseNo,
    name: 'contactDetails.licenseNo',
    elementType: 'input',
    onChange: (value: string) => {
      handleChange({ licenseNo: value });
    },
    onBlur: () => onBlurLicenseNumber(),
    validate: () => {
      if (loading) {
        return i18n('error.validatingLicenseNumber');
      }
      if (values.licenseValid === false) {
        return i18n('error.licenseNbrInvalid');
      }
      const typeValidation = checkValidationField(
        {
          type: validationTypes.CN_NUMBER,
        },
        values.licenseNo || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  // License field for other companies
  const contactLicenceNumberField = {
    label: i18n('field.comLicenseNumber'),
    value: values.contactLicenseNo,
    name: 'contactDetails.contactlicenseNo',
    elementType: 'input',
    onChange: (value: string) => {
      handleChange({ contactLicenseNo: value });
    },
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.contactLicenseNo || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const nationalityField = {
    label: i18n('field.nationality'),
    value: values.nationality,
    name: 'contactDetails.nationality',
    elementType: 'select',
    showSearch: companyType !== 'ADGE',
    items:
      companyType === 'ADGE'
        ? [
            {
              id: 'ARE',
              label: i18n('country.unitedarabemirates'),
            },
          ]
        : nationalities.map((item: IVariables) => {
            return {
              id: item.id,
              label: i18n(item.label),
            };
          }),
    onChange: (value: any) => handleChange({ nationality: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.nationality || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const mobileNoField = {
    label: i18n('field.mobilePhone'),
    value: values.mobileNo,
    elementType: 'inputTelephone',
    name: 'contactDetails.mobileNo',
    countries,
    onSelect: (value: any) => handleChange({ mobileNo: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.PHONE,
        },
        values.mobileNo || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const emailField = {
    label: i18n('applicantContact.email'),
    value: values.emailAddress,
    name: 'contactDetails.email',
    elementType: 'input',
    onChange: (value: any) => handleChange({ emailAddress: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.EMAIL,
        },
        values.emailAddress || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const companyNameENField = {
    label: i18n('field.companyNameEn'),
    value: values.englishName,
    name: 'contactDetails.englishName',
    elementType: 'input',
    onChange: (value: any) => handleChange({ englishName: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.englishName || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const companyNameARField = {
    label: i18n('field.companyNameAr'),
    value: values.arabicName,
    name: 'contactDetails.arabicName',
    elementType: 'input',
    onChange: (value: any) => handleChange({ arabicName: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.arabicName || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const legalFormField = {
    label: i18n('field.legalForm'),
    value: values.legalForm,
    name: 'contactDetails.legalForm',
    elementType: 'select',
    items: legalForms.map((item: IVariables) => {
      return {
        id: item.id,
        label: i18n(item.label),
      };
    }),
    onChange: (value: any) => handleChange({ legalForm: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.legalForm || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const repTypeField = {
    label: i18n('field.repTypeField'),
    value: values.representativeType,
    name: 'contactDetails.repType',
    elementType: 'select',
    items: representativeType.map((item: IVariables) => {
      return {
        id: item.id,
        label: i18n(item.label),
      };
    }),
    onChange: (value: any) =>
      handleRepresentativeTypeChange({
        representativeType: value,
      }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.representativeType || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  // Fully controlled component, must send 0 as initially value ( check initial state ) to control state strictly, sends back e
  const sharePercentageField = {
    label: i18n('field.sharePercentage'),
    disabled: () => {
      if (values.representativeType === '2') {
        return false;
      }
      return true;
    },
    precision: 2,
    step: 0.5,
    value: values.partnerSharePercentage,
    name: 'contactDetails.partnerSharePercentage',
    elementType: 'inputNumber',
    onChange: (e: any) => {
      const { value } = e.target;
      // This regex allows . but the final validation does not
      const partnerSharePercentageRegex = /^(100(?:\.00?)?|\d?\d(?:\.\d?\d?)?)$/;
      if (value.match(partnerSharePercentageRegex)) {
        handleChange({ partnerSharePercentage: String(value) });
      } else if (value === '') {
        handleChange({ partnerSharePercentage: String(value) });
      } else {
        handleChange({ partnerSharePercentage: '1' });
      }
    },
    validate: () => {
      if (values.representativeType === '2') {
        const typeValidation = validateSharePercentage(values, i18n);
        if (typeValidation) return typeValidation;
      }
      return undefined;
    },
  };

  const emirateField = {
    label: i18n('field.emirateOfBusiness'),
    value: values.emirate,
    name: 'contactDetails.emirate',
    elementType: 'select',
    items: emirates.map((item: IVariables) => {
      return {
        id: item.id,
        label: i18n(item.label),
      };
    }),
    onChange: (value: any) => handleChange({ emirate: value }),
    validate: () => {
      const typeValidation = checkValidationField(
        {
          type: validationTypes.REQUIRED,
        },
        values.emirate || null,
        startShowingErrors,
        i18n,
      );
      if (typeValidation) return typeValidation;
      return undefined;
    },
  };

  const allGccField = {
    label: i18n('field.allGcc'),
    name: 'contactDetails.allGcc',
    checked: values.allGcc,
    elementType: 'checkbox',
    onClick: (e: any) => {
      handleChange({ allGcc: e.target.checked });
    },
  };

  // These details may change with updated business requirements
  // Pass groups for formTemplate for each type of company
  const DEDCompanyTypeFields = [
    { twoColumns: true, fields: [licenceNumberField] },
  ];

  const ADGECompanyTypeFields = [
    {
      twoColumns: true,
      fields: [
        companyNameENField,
        companyNameARField,
        repTypeField,
        sharePercentageField,
        nationalityField,
      ],
    },
  ];

  const FZCompanyTypeFields = [
    {
      twoColumns: true,
      fields: [
        companyNameENField,
        companyNameARField,
        contactLicenceNumberField,
      ],
    },
    {
      twoColumns: true,
      fields: [
        repTypeField,
        sharePercentageField,
        nationalityField,
        mobileNoField,
        emailField,
      ],
    },
  ];

  const NLCompanyTypeFields = [
    {
      twoColumns: true,
      fields: [
        legalFormField,
        emirateField,
        companyNameENField,
        companyNameARField,
        contactLicenceNumberField,
      ],
    },
    {
      twoColumns: true,
      fields: [
        repTypeField,
        sharePercentageField,
        nationalityField,
        mobileNoField,
        emailField,
      ],
    },
  ];

  const FCCompanyTypeFields = [
    {
      twoColumns: true,
      fields: [
        companyNameENField,
        companyNameARField,
        contactLicenceNumberField,
      ],
    },
    {
      twoColumns: true,
      fields: [
        repTypeField,
        sharePercentageField,
        nationalityField,
        mobileNoField,
        emailField,
      ],
    },
    {
      twoColumns: false,
      fields: [allGccField],
    },
  ];

  const CompanyTypeFields: { [key: string]: any } = {
    DED: DEDCompanyTypeFields,
    ADGE: ADGECompanyTypeFields,
    FZ: FZCompanyTypeFields,
    NL: NLCompanyTypeFields,
    FC: FCCompanyTypeFields,
  };

  const companyTypeSelectField = [
    {
      'aria-label': 'select  ',
      label: i18n('field.companyType'),
      elementType: 'select',
      showSearch: false,
      name: 'field.companyType',
      value: companies[companyType].id,
      items: companyTypes.map((key: string) => {
        return {
          label: i18n(companies[key].label),
          id: companies[key].id,
        };
      }),
      onChange: (value: any) => handleCompanyTypeChange(value),
    },
  ];

  // Function to return fields based on companyType selected
  const getCompanyFormDetails = (fieldGroups: any[]): any[] => {
    if (companyTypes.length <= 1) {
      return [
        // companyType drop down group
        {
          fields: [],
          twoColumns: true,
          name: i18n('title.addCompanyDetails'),
        },
        // Field grups based on companyType selected
        ...fieldGroups.map((group: IVariables) => {
          const { fields } = group;
          return {
            twoColumns: group.twoColumns,
            // Unpack fields using methods
            fields,
          };
        }),
      ];
    }
    return [
      // companyType drop down group
      {
        fields: [...companyTypeSelectField],
        twoColumns: true,
        name: i18n('title.addCompanyDetails'),
      },
      // Field grups based on companyType selected
      ...fieldGroups.map((group: IVariables) => {
        const { fields } = group;
        return {
          twoColumns: group.twoColumns,
          // Unpack fields using methods
          fields,
        };
      }),
    ];
  };

  const formFieldGroups = CompanyTypeFields[companyType];
  const formGroups = getCompanyFormDetails(formFieldGroups);

  return (
    <React.Fragment>
      <FormTemplate inputGroups={formGroups} name="companyDetails" />
      <div style={margin} />
    </React.Fragment>
  );
};

export default CompanyDetailsQuestion;
