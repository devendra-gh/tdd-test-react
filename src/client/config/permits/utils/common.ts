import moment from 'moment';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const calculateSum = (items: any[], prop: string) => {
  const total = items.reduce((a, b) => {
    // eslint-disable-next-line security/detect-object-injection
    return a + parseFloat(b[prop]);
  }, 0);
  return total;
};

const formatDateForEstimatedFees = (date: string) => {
  return moment(date).format('MMM DD, YYYY');
};
const slugify = (text: string) => {
  return text
    .toLowerCase()
    .replace(/ /g, '-')
    .replace(/[^\w-]+/g, '');
};
export { calculateSum, formatDateForEstimatedFees, slugify };
