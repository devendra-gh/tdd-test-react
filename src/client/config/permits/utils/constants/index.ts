import { countries } from './countries';
import { legalForms } from './legalForms';
import { companies } from './companies';
import { emirates } from './emirates';
import { nationalities } from './nationalities';
import { permitType } from './permitType';
import { representativeType } from './representativeType';

export {
  companies,
  countries,
  emirates,
  legalForms,
  nationalities,
  permitType,
  representativeType,
};
