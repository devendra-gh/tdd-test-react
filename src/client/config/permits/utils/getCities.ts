const getCities = () => {
  return [
    {
      id: 'Abu Dhabi',
      label: 'global.AbuDhabi',
    },
    {
      id: 'Al Ain',
      label: 'global.AlAin',
    },
    {
      id: 'Al Dhafra',
      label: 'global.AlDhafra',
    },
  ];
};
export default getCities;
