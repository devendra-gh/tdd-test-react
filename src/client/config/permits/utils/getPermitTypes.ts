export const PERMIT_TYPE_ANNUAL = 'Annual';
export const PERMIT_TYPE_EVENT = 'Event';

export const DEFAULT_PERMIT_TYPE = PERMIT_TYPE_ANNUAL;

const permitTypes = {
  [PERMIT_TYPE_ANNUAL.toLowerCase()]: {
    id: PERMIT_TYPE_ANNUAL.toLowerCase(),
    label: 'global.annual',
    value: PERMIT_TYPE_ANNUAL,
  },
  [PERMIT_TYPE_EVENT.toLowerCase()]: {
    id: PERMIT_TYPE_EVENT.toLowerCase(),
    label: 'global.event',
    value: PERMIT_TYPE_EVENT,
  },
};
const getPermitTypes = () => {
  return Object.values(permitTypes);
};
export default getPermitTypes;
