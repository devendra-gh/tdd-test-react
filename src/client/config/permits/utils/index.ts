import getCities from './getCities';
import getPermitTypes from './getPermitTypes';
import getPlateCategories from './getPlateCategories';

export { getCities, getPermitTypes, getPlateCategories };
