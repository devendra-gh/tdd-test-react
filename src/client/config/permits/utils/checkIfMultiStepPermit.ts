import permitConfigs from 'client/config/permits/permitConfigs';
import { IVariables } from '@tamm/app-composer';
import { PERMIT_FOOD_TRUCK } from './constants/permits';
import { PERMIT_TYPE_EVENT } from './getPermitTypes';

const PermitConfigs: { [key: string]: any } = permitConfigs;

const checkIfMultiStepPermit = (props: IVariables) => {
  try {
    const { serviceType, permitInfo } = props;
    const serviceConfig = PermitConfigs[serviceType];
    if (serviceConfig) {
      const { permitDetails } = permitInfo[serviceType];
      if (
        serviceType === PERMIT_FOOD_TRUCK &&
        permitDetails.permitType === PERMIT_TYPE_EVENT
      ) {
        return false;
      }
      return serviceConfig.requiresEntityApproval;
    }
    return false;
  } catch (e) {
    return false;
  }
};
export default checkIfMultiStepPermit;
