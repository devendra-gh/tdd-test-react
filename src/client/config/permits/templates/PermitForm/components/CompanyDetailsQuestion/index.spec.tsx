import React from 'react';
import { shallow, mount } from 'enzyme';
import { checkValidationField } from 'client/config/utils/checkValidation';
import CompanyDetailsQuestion, { checkLicenseDetails } from './index';

jest.mock('client/config/utils/checkValidation', () => ({
  checkValidationField: jest.fn(),
  validateSharePercentage: jest.fn(),
  validationTypes: {
    CN_NUMBER: '',
  },
}));

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Company details question for type DED', () => {
  let props: any;

  beforeEach(() => {
    props = {
      values: {},
      serviceType: 'Air Ad',
      handleChange: jest.fn(),
      handleCompanyTypeChange: jest.fn(),
      handleRepresentativeTypeChange: jest.fn(),
      companyType: 'DED',
      i18n: (value: string) => value,
      fetchLicenseDetails: jest.fn().mockReturnValue({}).mockReturnValue({
        businessLicCity: 'Abu Dhabi',
        businessNameEng: 'Test Inc',
        businessNameArb: 'Test Inc Ar',
      }),
      companyTypes: ['DED'],
      startShowingErrors: false,
    };
  });

  it('should successfully render the company details form for type DED', () => {
    const component = shallow(<CompanyDetailsQuestion {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should return null when the service type is not provided', () => {
    props.serviceType = '';
    const component = shallow(<CompanyDetailsQuestion {...props} />);
    expect(component.type()).toEqual(null);
  });

  it('should fetch branch details from license number on blur', () => {
    props.values = { licenseNo: 'CN-1234567' };
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const licenseInput = component
      .find({ name: 'contactDetails.licenseNo' })
      .at(0);
    const allProps = licenseInput.props();
    const { onBlur } = allProps;
    onBlur();
    expect(props.fetchLicenseDetails).toHaveBeenCalled();
  });

  it('should not fetch branch details from empty license number on blur', () => {
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const licenseInput = component
      .find({ name: 'contactDetails.licenseNo' })
      .at(0);
    const allProps = licenseInput.props();
    const { onBlur } = allProps;
    onBlur();
    expect(props.fetchLicenseDetails).not.toHaveBeenCalled();
  });

  it('should call the callback function with the license details', async () => {
    const fetchLicenseDetails = jest.fn().mockReturnValue({
      businessLicCity: 'Abu Dhabi',
      businessNameEng: 'Test Inc',
      businessNameArb: 'Test Inc Ar',
    });
    const handleChange = jest.fn();
    const licenseNumber = 'CN-13444';
    await checkLicenseDetails(licenseNumber, fetchLicenseDetails, handleChange);
    expect(handleChange).toHaveBeenCalledWith({
      location: 'Abu Dhabi',
      businessNameEn: 'Test Inc',
      businessNameAr: 'Test Inc Ar',
      licenseValid: true,
    });
  });

  it('should call the callback function with the empty details on invalid license', async () => {
    const fetchLicenseDetails = jest.fn().mockReturnValue(undefined);
    const handleChange = jest.fn();
    const licenseNumber = 'CN-13444';
    await checkLicenseDetails(licenseNumber, fetchLicenseDetails, handleChange);
    expect(handleChange).toHaveBeenCalledWith({
      location: '',
      businessNameEn: '',
      businessNameAr: '',
      licenseValid: false,
    });
  });

  it('should use sane defaults for business details', async () => {
    let fetchLicenseDetails = jest.fn().mockReturnValue({
      businessLicCity: 'Abu Dhabi',
      businessNameArb: 'Test Inc Ar',
    });
    const handleChange = jest.fn();
    const licenseNumber = 'CN-13444';
    await checkLicenseDetails(licenseNumber, fetchLicenseDetails, handleChange);
    expect(handleChange).toHaveBeenCalledWith({
      location: 'Abu Dhabi',
      businessNameEn: 'Test Inc Ar',
      businessNameAr: 'Test Inc Ar',
      licenseValid: true,
    });
    fetchLicenseDetails = jest.fn().mockReturnValue({
      businessLicCity: 'Abu Dhabi',
      businessNameEng: 'Test Inc Ar',
    });
    await checkLicenseDetails(licenseNumber, fetchLicenseDetails, handleChange);
    expect(handleChange).toHaveBeenCalledWith({
      location: 'Abu Dhabi',
      businessNameEn: 'Test Inc Ar',
      businessNameAr: 'Test Inc Ar',
      licenseValid: true,
    });
    fetchLicenseDetails = jest.fn().mockReturnValue({});
    await checkLicenseDetails(licenseNumber, fetchLicenseDetails, handleChange);
    expect(handleChange).toHaveBeenCalledWith({
      location: '',
      businessNameEn: '',
      businessNameAr: '',
      licenseValid: true,
    });
  });
});

// eslint-disable-next-line no-console
console.log = () => {};
console.info = () => {};
console.warn = () => {};
console.error = () => {};

describe('Company details question for type other companies', () => {
  let props: any;

  beforeEach(() => {
    props = {
      values: {},
      serviceType: 'Air Ad',
      handleChange: jest.fn(),
      handleCompanyTypeChange: jest.fn(),
      handleRepresentativeTypeChange: jest.fn(),
      companyType: 'DED',
      i18n: (value: string) => value,
      fetchLicenseDetails: jest.fn().mockReturnValue({}).mockReturnValue({
        businessLicCity: 'Abu Dhabi',
        businessNameEng: 'Test Inc',
        businessNameArb: 'Test Inc Ar',
      }),
      companyTypes: ['DED', 'FC', 'FZ', 'NL', 'ADGE'],
      companyNameENField: {
        onChange: jest.fn(),
      },
    };
  });

  it('should successfully render the company details form for other companies', () => {
    const component = shallow(<CompanyDetailsQuestion {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should successfully render the company details form for type ADGE', () => {
    props.companyType = 'ADGE';
    const component = shallow(<CompanyDetailsQuestion {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('companyType radio group should work', () => {
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const companyTypeField = component
      .find({ name: 'field.companyType' })
      .at(0);
    const allProps = companyTypeField.props();
    const { onChange } = allProps;
    onChange({ target: { id: '0' } });
    expect(props.handleCompanyTypeChange).toHaveBeenCalled();
  });

  it('should call validate when licenseNo field is changed', () => {
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const licenseNo = component
      .find({ name: 'contactDetails.licenseNo' })
      .at(0);
    const allProps = licenseNo.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when contactlicenseNo field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const contactlicenseNo = component
      .find({ name: 'contactDetails.contactlicenseNo' })
      .at(0);
    const allProps = contactlicenseNo.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when nationality field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const nationality = component
      .find({ name: 'contactDetails.nationality' })
      .at(0);
    const allProps = nationality.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when mobileNo field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const mobileNo = component.find({ name: 'contactDetails.mobileNo' }).at(0);
    const allProps = mobileNo.props();
    const { validate, onSelect } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onSelect();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when email field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const email = component.find({ name: 'contactDetails.email' }).at(0);
    const allProps = email.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when englishName field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const englishName = component
      .find({ name: 'contactDetails.englishName' })
      .at(0);
    const allProps = englishName.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when arabicName field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const arabicName = component
      .find({ name: 'contactDetails.arabicName' })
      .at(0);
    const allProps = arabicName.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when legalForm field is changed', () => {
    props.companyType = 'NL';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const legalForm = component
      .find({ name: 'contactDetails.legalForm' })
      .at(0);
    const allProps = legalForm.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when repType field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const repType = component.find({ name: 'contactDetails.repType' }).at(0);
    const allProps = repType.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleRepresentativeTypeChange).toHaveBeenCalled();
  });

  it('should call validate when sharePercentage field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const sharePercentage = component
      .find({ name: 'contactDetails.partnerSharePercentage' })
      .at(0);
    const allProps = sharePercentage.props();
    const { validate, onChange, disabled } = allProps;
    expect(disabled()).toBe(true);
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange({ target: { value: '49' } });
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('sharePercentage field should be enabled if repType is 2', () => {
    props.companyType = 'FC';
    props.values = { representativeType: '2' };
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const sharePercentage = component
      .find({ name: 'contactDetails.partnerSharePercentage' })
      .at(0);
    const allProps = sharePercentage.props();
    const { disabled } = allProps;
    expect(disabled()).toBe(false);
  });

  it('should call validate when emirate field is changed', () => {
    props.companyType = 'NL';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const emirate = component.find({ name: 'contactDetails.emirate' }).at(0);
    const allProps = emirate.props();
    const { validate, onChange } = allProps;
    validate();
    expect(checkValidationField).toHaveBeenCalled();
    onChange();
    expect(props.handleChange).toHaveBeenCalled();
  });

  it('should call validate when allGcc field is changed', () => {
    props.companyType = 'FC';
    const component = mount(<CompanyDetailsQuestion {...props} />);
    const allGcc = component.find({ name: 'contactDetails.allGcc' }).at(0);
    const allProps = allGcc.props();
    const { onClick } = allProps;
    onClick({ target: { checked: false } });
    expect(props.handleChange).toHaveBeenCalled();
  });
});
