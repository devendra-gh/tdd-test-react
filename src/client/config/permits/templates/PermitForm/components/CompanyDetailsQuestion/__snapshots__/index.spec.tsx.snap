// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company details question for type DED should successfully render the company details form for type DED 1`] = `
<Fragment>
  <LocalizedComponent(FormTemplate)
    cancelButton={null}
    cancelLink={null}
    formData={Object {}}
    i18n={[Function]}
    id={null}
    inputGroups={
      Array [
        Object {
          "fields": Array [],
          "name": "title.addCompanyDetails",
          "twoColumns": true,
        },
        Object {
          "fields": Array [
            Object {
              "elementType": "input",
              "label": "field.comLicenseNumber",
              "name": "contactDetails.licenseNo",
              "onBlur": [Function],
              "onChange": [Function],
              "validate": [Function],
              "value": undefined,
            },
          ],
          "twoColumns": true,
        },
      ]
    }
    name="companyDetails"
    submitButton={null}
  />
  <div
    style={
      Object {
        "margin": "20px",
      }
    }
  />
</Fragment>
`;

exports[`Company details question for type other companies should successfully render the company details form for other companies 1`] = `
<Fragment>
  <LocalizedComponent(FormTemplate)
    cancelButton={null}
    cancelLink={null}
    formData={Object {}}
    i18n={[Function]}
    id={null}
    inputGroups={
      Array [
        Object {
          "fields": Array [
            Object {
              "aria-label": "select  ",
              "elementType": "select",
              "items": Array [
                Object {
                  "id": "DED",
                  "label": "company.label.ded",
                },
                Object {
                  "id": "FC",
                  "label": "company.label.fc",
                },
                Object {
                  "id": "FZ",
                  "label": "company.label.fz",
                },
                Object {
                  "id": "NL",
                  "label": "company.label.nl",
                },
                Object {
                  "id": "ADGE",
                  "label": "company.label.adge",
                },
              ],
              "label": "field.companyType",
              "name": "field.companyType",
              "onChange": [Function],
              "showSearch": false,
              "value": "DED",
            },
          ],
          "name": "title.addCompanyDetails",
          "twoColumns": true,
        },
        Object {
          "fields": Array [
            Object {
              "elementType": "input",
              "label": "field.comLicenseNumber",
              "name": "contactDetails.licenseNo",
              "onBlur": [Function],
              "onChange": [Function],
              "validate": [Function],
              "value": undefined,
            },
          ],
          "twoColumns": true,
        },
      ]
    }
    name="companyDetails"
    submitButton={null}
  />
  <div
    style={
      Object {
        "margin": "20px",
      }
    }
  />
</Fragment>
`;

exports[`Company details question for type other companies should successfully render the company details form for type ADGE 1`] = `
<Fragment>
  <LocalizedComponent(FormTemplate)
    cancelButton={null}
    cancelLink={null}
    formData={Object {}}
    i18n={[Function]}
    id={null}
    inputGroups={
      Array [
        Object {
          "fields": Array [
            Object {
              "aria-label": "select  ",
              "elementType": "select",
              "items": Array [
                Object {
                  "id": "DED",
                  "label": "company.label.ded",
                },
                Object {
                  "id": "FC",
                  "label": "company.label.fc",
                },
                Object {
                  "id": "FZ",
                  "label": "company.label.fz",
                },
                Object {
                  "id": "NL",
                  "label": "company.label.nl",
                },
                Object {
                  "id": "ADGE",
                  "label": "company.label.adge",
                },
              ],
              "label": "field.companyType",
              "name": "field.companyType",
              "onChange": [Function],
              "showSearch": false,
              "value": "ADGE",
            },
          ],
          "name": "title.addCompanyDetails",
          "twoColumns": true,
        },
        Object {
          "fields": Array [
            Object {
              "elementType": "input",
              "label": "field.companyNameEn",
              "name": "contactDetails.englishName",
              "onChange": [Function],
              "validate": [Function],
              "value": undefined,
            },
            Object {
              "elementType": "input",
              "label": "field.companyNameAr",
              "name": "contactDetails.arabicName",
              "onChange": [Function],
              "validate": [Function],
              "value": undefined,
            },
            Object {
              "elementType": "select",
              "items": Array [
                Object {
                  "id": "1",
                  "label": "repType.owner",
                },
                Object {
                  "id": "2",
                  "label": "repType.partner",
                },
                Object {
                  "id": "3",
                  "label": "repType.sponsor",
                },
                Object {
                  "id": "4",
                  "label": "repType.manager",
                },
                Object {
                  "id": "5",
                  "label": "repType.heirsRepresentative",
                },
                Object {
                  "id": "7",
                  "label": "repType.commisionerToSign",
                },
              ],
              "label": "field.repTypeField",
              "name": "contactDetails.repType",
              "onChange": [Function],
              "validate": [Function],
              "value": undefined,
            },
            Object {
              "disabled": [Function],
              "elementType": "inputNumber",
              "label": "field.sharePercentage",
              "name": "contactDetails.partnerSharePercentage",
              "onChange": [Function],
              "precision": 2,
              "step": 0.5,
              "validate": [Function],
              "value": undefined,
            },
            Object {
              "elementType": "select",
              "items": Array [
                Object {
                  "id": "ARE",
                  "label": "country.unitedarabemirates",
                },
              ],
              "label": "field.nationality",
              "name": "contactDetails.nationality",
              "onChange": [Function],
              "showSearch": false,
              "validate": [Function],
              "value": undefined,
            },
          ],
          "twoColumns": true,
        },
      ]
    }
    name="companyDetails"
    submitButton={null}
  />
  <div
    style={
      Object {
        "margin": "20px",
      }
    }
  />
</Fragment>
`;
